src/libraries/System.Reflection.Emit/src/System/Reflection/Emit/ModuleBuilderImpl.cs

```diff
@@ -0,0 +1,125 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using System.Reflection.Metadata;
+using System.Reflection.Metadata.Ecma335;
+using System.Runtime.InteropServices;
+
+namespace System.Reflection.Emit
+{
+    internal sealed class ModuleBuilderImpl : ModuleBuilder
+    {
+        private readonly AssemblyBuilderImpl _assemblyBuilder;
+
+        #region Internal Data Members
+
+        internal Dictionary<Assembly, AssemblyReferenceHandle> _assemblyRefStore = new Dictionary<Assembly, AssemblyReferenceHandle>();
+        internal Dictionary<Type, TypeReferenceHandle> _typeRefStore = new Dictionary<Type, TypeReferenceHandle>();
+        internal List<TypeBuilderImpl> _typeDefStore = new List<TypeBuilderImpl>();
+        internal int _nextMethodDefRowId = 1;
+        internal int _nextFieldDefRowId = 1;
+        internal const string ManifestModuleName = "RefEmit_InMemoryManifestModule";
+
+        #endregion
+
+
+        internal ModuleBuilderImpl(string name, AssemblyBuilderImpl assembly)
+        {
+            _assemblyBuilder = assembly;
+            ScopeName = name;
+        }
+
+        internal void AppendMetadata(MetadataBuilder metadata)
+        {
+            // Add module metadata
+            metadata.AddModule(
+                generation: 0,
+                metadata.GetOrAddString(ScopeName),
+                metadata.GetOrAddGuid(Guid.NewGuid()),
+                default,
+                default);
+
+            // Create type definition for the special <Module> type that holds global functions
+            metadata.AddTypeDefinition(
+                default,
+                default,
+                metadata.GetOrAddString("<Module>"),
+                baseType: default,
+                fieldList: MetadataTokens.FieldDefinitionHandle(1),
+                methodList: MetadataTokens.MethodDefinitionHandle(1));
+
+            // Add each type definition to metadata table.
+            foreach (TypeBuilderImpl typeBuilder in _typeDefStore)
+            {
+                TypeReferenceHandle parent = default;
+                if (typeBuilder.BaseType is not null)
+                {
+                    parent = GetTypeReference(metadata, typeBuilder.BaseType);
```

Comment:
This should emit type definition when the base type is from the same assembly.

Reply:
Right, that will be added later

Reply:
Add TODO comment?